tests/openassetio-core/trait/TraitBaseTest.cpp

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2022 The Foundry Visionmongers Ltd
+#include <openassetio/trait/TraitBase.hpp>
+#include <openassetio/typedefs.hpp>
+
+#include <catch2/catch.hpp>
+#include "openassetio/TraitsData.hpp"
+
+namespace {
+
+using openassetio::Str;
+using openassetio::TraitsData;
+using openassetio::TraitsDataPtr;
+using openassetio::trait::TraitBase;
+using openassetio::trait::TraitId;
+using openassetio::trait::TraitPropertyStatus;
+using openassetio::trait::property::Key;
+
+// TraitBase can't be tested directly, so we derive a test trait.
+struct TestTrait : TraitBase<TestTrait> {
+ private:
+  static inline const Key kSomeProperty{"some property"};
+
+ public:
+  static inline const TraitId kId{"test"};
+
+  using TraitBase<TestTrait>::TraitBase;
+
+  [[nodiscard]] TraitPropertyStatus getSomeProperty(Str* out) const {
+    return getTraitProperty(out, kId, kSomeProperty);
+  }
+
+  void setSomeProperty(const Str& value) { data()->setTraitProperty(kId, kSomeProperty, value); }
+
+  // Helpers for tests
+
+  [[nodiscard]] TraitsDataPtr& dataPtr() { return data(); }
+  [[nodiscard]] const TraitsDataPtr& dataPtr() const { return data(); }
+};
+
+SCENARIO("Retrieving the undelying data") {
+  GIVEN("Some known traits data") {
+    TraitsDataPtr data = std::make_shared<TraitsData>();
+
+    WHEN("a trait instance is constructed with data") {
+      TestTrait trait(data);
+
+      THEN("the supplied data is exposed via protected method") { CHECK(trait.dataPtr() == data); }
+
+      WHEN("a const trait instance is constructed with the data") {
+        const TestTrait constTrait(data);
+
+        THEN("the supplied data is exposed via const protected method") {
+          CHECK(constTrait.dataPtr() == data);
+        }
+      }
+    }
+  }
+}
+
+SCENARIO("Checking a trait is valid") {
+  GIVEN("Some known traits data") {
+    TraitsDataPtr data = std::make_shared<TraitsData>();
+
+    AND_GIVEN("the data has the trait set") {
+      data->addTrait(TestTrait::kId);
+
+      WHEN("called") {
+        TestTrait trait(data);
+        THEN("then isValid returns true") { CHECK(trait.isValid()); }
+      }
+    }
+
+    AND_GIVEN("the data does not have trait") {
+      TestTrait trait(data);
+      THEN("the isValid returns false") { CHECK(!trait.isValid()); }
+    }
+  }
+}
+
+SCENARIO("Imbuing a trait to the traits data held by a trait instance") {
+  GIVEN("Some known traits data held by a trait") {
+    TraitsDataPtr data = std::make_shared<TraitsData>();
+    TestTrait trait(data);
+
+    AND_GIVEN("the data does not have the trait set") {
+      WHEN("the trait is imbued") {
+        trait.imbue();
+        THEN("trait is added") { CHECK(data->hasTrait(TestTrait::kId)); }
+      }
+    }
+
+    AND_GIVEN("the data does have the trait set") {
+      data->addTrait(TestTrait::kId);
+
+      WHEN("the trait is imbued") {
+        THEN("is a noop") { trait.imbue(); }
```

Comment:
[consider] I wonder if we can do a little better than no assertions.  We could copy-construct the data before `imbue`ing and check equality after?

Reply:
Improved to match python checks [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/373/commits/8a2fabac6bb57da2c4bd23a2e09f993a29ceb818).